src/frontend/src/optimizer/plan_node/logical_join.rs

```diff
@@ -842,6 +912,56 @@ impl PredicatePushdown for LogicalJoin {
         let left_predicate = left_from_filter.and(left_from_on);
         let right_predicate = right_from_filter.and(right_from_on);
 
+        // Derive conditions to push to the other side based on eq condition columns
+        let left_cols_num = self.left().schema().len();
+        let right_cols_num = self.right().schema().len();
+        let eq_condition = EqJoinPredicate::create(left_cols_num, right_cols_num, new_on.clone());
+
+        // Only push to RHS if RHS is inner side of a join (RHS requires match on LHS)
+        // TODO: make this a function
+        let right_derived_predicate = if matches!(
+            join_type,
+            JoinType::Inner | JoinType::LeftOuter | JoinType::RightSemi
```

Comment:
It seems `LeftSemiJoin` and `RightSemiJoin` are both acceptable here.

Reply:
For the `LeftSemi` case, indeed, we will never need a row on RHS if matching rows on LHS are filtered out.